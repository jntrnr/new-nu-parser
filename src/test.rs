use crate::resolver::Resolver;
use crate::{compiler::Compiler, parser::Parser};
use std::path::Path;

fn evaluate_example(fname: &Path) -> String {
    let mut compiler = Compiler::new();
    let contents = std::fs::read(fname).expect("We only run tests found by glob");

    // Homogenize line endings from CRLF to LF to make sure tests work both on Windows and Linux/macOS
    let contents = String::from_utf8(contents)
        .unwrap()
        .replace("\r\n", "\n")
        .as_bytes()
        .to_owned();

    let span_offset = compiler.span_offset();
    compiler.add_file(&fname.to_string_lossy(), &contents);

    let parser = Parser::new(compiler, span_offset);
    compiler = parser.parse();

    let mut result = compiler.display_state();

    if !compiler.errors.is_empty() {
        return result;
    }

    let mut resolver = Resolver::new(&compiler);
    resolver.resolve();
    result.push_str(&resolver.display_state());

    compiler.merge_name_bindings(resolver.to_name_bindings());

    result
}

#[test]
fn test_node_output() {
    insta::glob!("../tests", "*.nu", |path| {
        insta::assert_snapshot!(evaluate_example(path));
    });
}
