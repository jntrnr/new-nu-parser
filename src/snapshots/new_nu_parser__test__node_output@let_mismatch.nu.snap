---
source: src/test.rs
expression: evaluate_example(path)
input_file: tests/let_mismatch.nu
---
==== COMPILER ====
0: Variable (4 to 5)
1: Name (7 to 13)
2: Type { name: NodeId(1), params: None, optional: false } (7 to 13)
3: Int (16 to 19)
4: Let { variable_name: NodeId(0), ty: Some(NodeId(2)), initializer: NodeId(3), is_mutable: false } (0 to 19)
5: Variable (21 to 23)
6: Block(BlockId(0)) (0 to 24)
==== SCOPE ====
0: Frame Scope, variables: [ x: NodeId(0) ], node_id: NodeId(6)
==== TYPES ====
0: string
1: unknown
2: string
3: int
4: ()
5: string
6: ()
==== TYPE ERRORS ====
Error (NodeId 3): initializer does not match declared type

